core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java

```diff
@@ -149,36 +152,41 @@ public void onNext(HttpObject o) {
 
                 final ResponseHeaders headers = (ResponseHeaders) o;
                 final HttpStatus status = headers.status();
-                final ResponseHeaders merged;
                 if (status.isInformational()) {
                     if (endOfStream) {
                         req.abortResponse(new IllegalStateException(
                                 "published an informational headers whose endOfStream is true: " + o +
                                 " (service: " + service() + ')'), true);
                         return;
                     }
-                    merged = headers;
-                } else {
-                    if (responseEncoder.isResponseHeadersSent(req.id(), req.streamId())) {
-                        // The response is sent by the HttpRequestDecoder so we just cancel the stream message.
-                        isComplete = true;
-                        setDone(true);
-                        return;
+                    if (!isHttp1() || !isHttp1WebSocketUpgradeResponse(headers)) {
```

Comment:
Would it be more readable if we use parens?

```suggestion
                    if (!(isHttp1() && isHttp1WebSocketUpgradeResponse(headers))) {
```